models/percona_sso_model_helpers.go

```diff
@@ -17,32 +17,128 @@
 	"gopkg.in/reform.v1"
 )
 
+var perconaSSOMtx sync.Mutex
+
 // GetPerconaSSODetails returns PerconaSSODetails if there are any, error otherwise.
-func GetPerconaSSODetails(q *reform.Querier) (*PerconaSSODetails, error) {
-	ssoDetails, err := q.SelectOneFrom(PerconaSSODetailsView, "")
+// Access token is automatically refreshed if it is expired.
+// Get, check eventually refresh done in one tx.
+func GetPerconaSSODetails(ctx context.Context, q *reform.Querier) (*PerconaSSODetails, error) {
+	perconaSSOMtx.Lock()
+	defer perconaSSOMtx.Unlock()
+
+	ssoDetails, err := q.SelectOneFrom(PerconaSSODetailsTable, "")
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to get Percona SSO Details")
 	}
-	return ssoDetails.(*PerconaSSODetails), nil
+
+	details := ssoDetails.(*PerconaSSODetails)
+	if details.isAccessTokenExpired() {
+		refreshedToken, err := details.refreshAndGetAccessToken(ctx, q)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to insert Percona SSO Details")
+		}
+		details.AccessToken = refreshedToken
+	}
+
+	return details, nil
+}
+
+// GetPerconaSSOAccessToken returns PerconaSSOAccessToken if there is GetPerconaSSODetails, error otherwise.
+func GetPerconaSSOAccessToken(ctx context.Context, q *reform.Querier) (*PerconaSSOAccessToken, error) {
+	ssoDetails, err := GetPerconaSSODetails(ctx, q)
+	if err != nil {
+		return nil, err
+	}
+
+	return ssoDetails.AccessToken, nil
+}
+
+func (sso *PerconaSSODetails) refreshAndGetAccessToken(ctx context.Context, q *reform.Querier) (*PerconaSSOAccessToken, error) {
+	url := fmt.Sprintf("%s%s", sso.IssuerURL, sso.Scope)
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	authHeader := base64.StdEncoding.EncodeToString([]byte(sso.ClientID + ":" + sso.ClientSecret))
+	h := req.Header
+	h.Add("Authorization", "Basic "+authHeader)
+	h.Add("Accept", "application/json")
+	h.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	timeBeforeRequest := time.Now()
+	client := &http.Client{}
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close() //nolint:errcheck
+
+	if res.StatusCode == http.StatusOK {
+		var accessToken *PerconaSSOAccessToken
+		bodyBytes, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			return nil, err
+		}
+		if err := json.Unmarshal(bodyBytes, &accessToken); err != nil {
+			return nil, err
+		}
+
+		accessToken.ExpiresAt = timeBeforeRequest.Add(time.Duration(accessToken.ExpiresIn) * time.Second)
+		sso.AccessToken = accessToken
+
+		if err := q.UpdateColumns(sso, "access_token"); err != nil {
+			return nil, err
+		}
+
+		return accessToken, nil
+	}
+
+	return nil, fmt.Errorf("get access token of Percona SSO Details failed, status code: %d", res.StatusCode)
+}
+
+func (sso *PerconaSSODetails) isAccessTokenExpired() bool {
+	if sso == nil || sso.AccessToken == nil {
+		return true
+	}
+
+	return time.Now().After(sso.AccessToken.ExpiresAt)
```

Comment:
I would add one more condition:
```suggestion
	return time.Now().After(sso.AccessToken.ExpiresAt.Add(-Minute * 5))
```

to reserve some window to perform API call to Portal (and verify this token on Portal side).

So that the logic is the following: refresh access token BEFORE it actually expires.